Source/FieldSolver/ElectrostaticSolver.cpp

```diff
@@ -346,17 +370,126 @@ WarpX::computePhiCartesian (const amrex::Vector<std::unique_ptr<amrex::MultiFab>
 #endif
+        for ( MFIter mfi(*phi[lev], TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+            // Extract the potential
+            auto phi_arr   = phi[lev]->array(mfi);
+            // Extract tileboxes for which to loop
+            const Box& tb  = mfi.tilebox( phi[lev]->ixType().toIntVect() );
+
+            amrex::ParallelFor( tb,
+                [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+                    int idx, idx_lo, idx_hi;
+
+                    for (int idim=0; idim<AMREX_SPACEDIM; idim++){
+                        // check if the boundary in this dimension should be set
+                        if (!dirichlet_flag[idim]) continue;
+
+                        if (idim == 0){
+                            idx = i;
+                            idx_lo = lo.x;
+                            idx_hi = hi.x;
+                        }
+                        if (idim == 1){
+                            idx = j;
+                            idx_lo = lo.y;
+                            idx_hi = hi.y;
+                        }
+                        if (idim == 2){
+                            idx = k;
+                            idx_lo = lo.z;
+                            idx_hi = hi.z;
+                        }
+
+                        if (idx == idx_lo) phi_arr(i,j,k) = phi_bc_values_lo[idim];
+                        if (idx == idx_hi) phi_arr(i,j,k) = phi_bc_values_hi[idim];
+                    }
+
+                } // loop ijk
+            );
+    }} // lev & MFIter
+}
+
+/* \bried Utility function to parse input file for boundary potentials.
+
+    The input values are parsed to allow math expressions for the potentials
+    that specify time dependence.
+
+   \param[in] idim The dimension for which the potential is queried
+   \param[inout] pot_lo The specified value of `phi` on the lower boundary.
+   \param[inout] pot_hi The specified value of `phi` on the upper boundary.
+*/
+void
+WarpX::getPhiBC( const int idim, amrex::Real &pot_lo, amrex::Real &pot_hi ) const
+{
+    Vector<std::string> potential_lo_str(AMREX_SPACEDIM);
+    Vector<std::string> potential_hi_str(AMREX_SPACEDIM);
+
+    // set default potentials to zero in order for current tests to pass
+    // but forcing the user to specify a potential might be better
+    for (int i = 0; i < AMREX_SPACEDIM; i++)
+    {
+        potential_lo_str[i] = "0";
+        potential_hi_str[i] = "0";
+    }
+
+    // Get the boundary potentials specified in the simulation input file
+    // first as strings and then parse those for possible math expressions
+    ParmParse pp_geom("boundary");
+    pp_geom.queryarr("potential_lo",potential_lo_str,0,AMREX_SPACEDIM);
+    pp_geom.queryarr("potential_hi",potential_hi_str,0,AMREX_SPACEDIM);
+
+    auto parser_lo = makeParser(potential_lo_str[idim], {"t"});
+    pot_lo = parser_lo.eval(gett_new(0));
+    auto parser_hi = makeParser(potential_hi_str[idim], {"t"});
+    pot_hi = parser_hi.eval(gett_new(0));
+}
```

Comment:
I'm not a fan of re-parsing the expressions at every solve invocation, rather than parsing them once and then just evaluating. Without profiling saying this is slow, I don't see a need to hold things up over it, though.

Reply:
I agree and I think this is another example where it would be good to have a persistent field solver object in which the parser can be kept along with most things in `AddSpaceChargeFieldLabFrame` and the boundary condition specifications. This will also allow for an easy way to turn off the B field solve and not overwriting `phi` at the start of every solve. But I think all that should go in a different PR.